src/provision.toit

```diff
@@ -0,0 +1,494 @@
+      r.read_field SET_CONFIG_SSID:
+        ssid = r.read_primitive protobuf.PROTOBUF_TYPE_STRING
+      r.read_field SET_CONFIG_PASSWORD:
+        password = r.read_primitive protobuf.PROTOBUF_TYPE_STRING
+    
+    resp_msg := {
+      MSG: MSG_RESP_CONFIG
+    }
+    resp := protobuf_map_to_bytes_ --message=resp_msg
+    return resp
+
+  set_apply -> ByteArray:
+    network = wifi.open
+        --ssid=ssid
+        --password=password
+
+    resp_msg := {
+        MSG: MSG_RESP_APPY
+    }
+    resp := protobuf_map_to_bytes_ --message=resp_msg
+    return resp   
+
+  req_status -> ByteArray:
+    ap ::= network.access_point
+
+    resp_msg := {
+      MSG: MSG_RESP_STATUS,
+      RESP_STATUS: {
+        RESP_STATUS_CONNECTED: {
+          RESP_STATUS_CONNECTED_IPV4_ADDR: "$(network.address)",
+          RESP_STATUS_CONNECTED_AUTH_MODE: ap.authmode,
+          RESP_STATUS_CONNECTED_SSID: ap.ssid,
+          RESP_STATUS_CONNECTED_BSSID: ap.bssid,
+          RESP_STATUS_CONNECTED_CHANNEL: ap.channel,
+        }
+      }
+    }
+    resp := protobuf_map_to_bytes_ --message=resp_msg
+    return resp     
+
+  run data/ByteArray -> ByteArray:
+    resp := #[]
+
+    r := protobuf.Reader data
+    r.read_message:
+      r.read_field MSG:
+        msgid := r.read_primitive protobuf.PROTOBUF_TYPE_INT32
+        if msgid == MSG_SET_APPLY:
+          resp = set_apply
+      r.read_field SET_CONFIG:
+        resp = set_config r
+      r.read_field REQ_STATUS:
+        tmp := r.read_primitive protobuf.PROTOBUF_TYPE_BYTES
+        resp = req_status
+        latch.set true
+      r.read_field REQ_APPLY:
+        tmp := r.read_primitive protobuf.PROTOBUF_TYPE_BYTES
+        resp = set_apply
+
+    return resp
+
+class Provision:
+  service_/BLEService_ := ?
+  security_/Security_ := ?
+  latch_ := monitor.Latch
+
+  constructor.ble  service_name/string security/Security_:
+    return Provision.ble_with_uuid SERVICE_UUID service_name security
+  
+  constructor.ble_with_uuid service_uuid/ByteArray service_name/string .security_/Security_:
+    service_ = BLEService_ service_uuid service_name
+
+  start -> none:
+    task:: ch_ver_task_
+    task:: ch_config_task_
+    task:: ch_session_task_
+    task:: ch_scan_task_
+
+    service_.start
+  
+  wait -> none:
+    ret := latch_.get
+
+  static common_process_ security/Security_ process/Process_ characteristic/BLECharacteristic_:
+    encrypt_data := characteristic.read
+    data := security.decrypt encrypt_data
+    resp := process.run data
+    if resp.size > 0:
+      data = security.encrypt resp
+      characteristic.write data
+
+  ch_ver_task_:
+    characteristic := service_["proto-ver"]
+    session_process := VerProcess_ security_.version
+    common_process_ security_ session_process characteristic
+
+  ch_session_task_:
+    characteristic := service_["prov-session"]
+    session_process := SessionProcess_
+    while true:
+      common_process_ security_ session_process characteristic
```

Comment:
When is this task (and the scan/config task) canceled?
Are they not running indefinitely?

Reply:
These tasks should be kept until the BLE provisioning really ends, I think there are 2 method to mark this ending:
1. users disable BLE provisioning
2. the class `Provision` de-structure
In other case, these tasks should be keep running.

Reply:
That sounds like there are three possible places where we want to cancel the tasks:
1. the provisioning is successful
2. the provisioning is canceled by the user (by some call to the class)
3. the provisioning instance is freed by the garbage collector

In all three cases, we need to store the created task in a field in the object:
```
// Add fields `version_task_/Task? := null` ... to the class.
 
start -> none:
  if version_task_: throw "Already running".
  version_task_ = task:: ch_ver_task_
  config_task_ = task:: ch_config_task_
 ...
```

The first could maybe be done inside the `ch_config_task_`. Once it receives a valid config, it could shut down the tasks.

The second could be something like:
```
  /** Closes the provisioning and shuts down the service. */
  close:
    if version_task_:
      version_task_.cancel
      version_task_ = null
    if config_task_:
      config_task_.cancel
      config_task_ = null
```
It's probably a good idea to also fill the latch in that case. So maybe the last line of `close` could be:
```
    if not latch_.has_value: latch_.set null
```
or, if we prefer,
```
    if not latch_.has_value: latch_.set --exception "Provisioning interrupted"
```
We should then document what happens for the `wait` function.

The third would require a bit more work: we would need to add a finalizer somewhere, so that everything is shut down when the object dies. However, the tasks have a reference to it, so we wouldn't be able to add finalizers to the `Provision_` object, as the tasks would keep it alive. We could wrap the `Provision_` object and put a finalizer on the wrapping object, which then calls `close`. For now I would say that users either have to complete the provisioning successfully, or call `close` (or even call `close` when things are successful).

Reply:
I plan to put 4 tasks' closing in the function close. this may look easier and more readable.
Modify the function `wait` return type from `none` to `bool`, then
1. Provisioning successfully:
```
latch_.set true
```
2. Provisioning is uncompleted and close
```
latch_.set false
```
User can check the result of function `wait` and know what happened.

Reply:
About the `add finalizers`, I check some toit code, and find following usage
```
class AAAA
    constructor:
       add_finalizer this::
         this.close
```

Reply:
You can't just use a finalizer on the `Provision` object, as the tasks have a handle to it, and thus keep it alive. Remember: the finalizer only gets called if an object is garbage collected, and that only happens if no other code has a reference to it. However, the tasks would have an (implicit "this") reference to it.

You could create a two-layer approach:
```
class Provision:
  nested_provision_/Provision_
  ...

  constructor.ble service_name/string security/Security_:
    return Provision.ble_with_uuid SERVICE_UUID service_name security

  constructor.ble_with_uuid service_uuid/ByteArray service_name/string .security_/Security_:
    nested_provision_ = Provision_.ble_with_uuid ...
    add_finalizer this::
      this.close

  close:
    nested_provision_.close

  // Forward all methods.
  start:
    nested_provision_.start
```
In that case, the nested provision would be shut down as soon as the `Provision` object is gone.
I would suggest doing this in a separate PR.

Reply:
I have not understand toit task, I have following questions:

1. Except task automatically exits from it self's logic,  task will not be cancelled  until we call function like `task_handle.cancel` ?
2. because task is not cancelled, so this cause class which owns this task also to be not collected?
3. We use a parent class can avoid issue of upper 2 ?

Reply:
There are two kinds of tasks: "normal" and "background".
A normal task keeps running until the lambda returns, or until something calls `cancel` on it.
A background task does the same, but also stops running if its parent task stops running.

An example with a "normal" task:
```
// This program will not stop.
main:
  task::
    while true:
      // Runs forever.
      sleep --ms=1000
```

An example with a "background" task:
```
main:
  // Stops when the parent task dies.
  task --background:
    while true:
      sleep --ms=1000
```

In our example, the lambda that is used to start the task has a reference to the instance that contains the task objects. There are a few places where `this` is implicitly captured.:
```
    version_task_ = task:: ch_version_task_
    config_task_ = task:: ch_config_task_
    session_task_ = task:: ch_session_task_
    scan_task_ = task:: ch_scan_task_
```
In all these cases, the lambda implicitly captures `this`:
```
    version_task_ = task:: this.ch_version_task_
    config_task_ = task:: this.ch_config_task_
    session_task_ = task:: this.ch_session_task_
    scan_task_ = task:: this.ch_scan_task_
```
This means, that each of these lambda objects has a field with `this` in it.

Similarly, the `this` object is again used in the `ch_config_task_` method:
```
   session_process := ConfigProcess_ latch_
```
is equivalent to:
```
   session_process := ConfigProcess_ this.latch_
```

So all these lambdas have a pointer to the object, and the garbage collector thus isn't allowed to free it. Since the object isn't freed, the finalizer is never called.